File: src/javalang/array/BinaryTreeLevelOrderTraversal.java
package javalang.array;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class BinaryTreeLevelOrderTraversal {

    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> result1 = new ArrayList<>();
            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll();
                if (node != null) {
                    result1.add(node.val);
                    queue.add(node.left);
                    queue.add(node.right);
                }
            }
            if (!result1.isEmpty())
                result.add(result1);
        }
        return result;
    }

    //test  code
    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);
        root.right.left = new TreeNode(6);
        root.right.right = new TreeNode(7);
        root.left.left.left = new TreeNode(8);
        root.left.left.right = new TreeNode(9);
        root.left.right.left = new TreeNode(10);
        root.left.right.right = new TreeNode(11);
        root.right.left.left = new TreeNode(12);
        root.right.left.right = new TreeNode(13);
        root.right.right.left = new TreeNode(14);
        root.right.right.right = new TreeNode(15);
        BinaryTreeLevelOrderTraversal bt = new BinaryTreeLevelOrderTraversal();
        List<List<Integer>> result = bt.levelOrder(root);
        for (List<Integer> list : result) {
            System.out.println(list);
        }
    }
}

File: src/kotlin/tree/BinaryTreeLevelOrderTraversal.kt
import java.util.*
import kotlin.collections.ArrayList

class BinaryTreeLevelOrderTraversal {

    data class TreeNode(val `val`: Int, var left: TreeNode? = null, var right: TreeNode? = null) {
        override fun toString(): String {
            return `val`.toString()
        }
    }


    fun levelOrder(root: TreeNode?): List<List<Int>> {
        if (root == null) {
            return listOf()
        }
        val result = mutableListOf<List<Int>>()
        val queue = mutableListOf<TreeNode>()
        queue.add(root)
        while (queue.isNotEmpty()) {
            val size = queue.size
            val level = mutableListOf<Int>()
            for (i in 0 until size) {
                val node: TreeNode = queue.removeAt(0)
                level.add(node.`val`)
                if (node.left != null) {
                    queue.add(node.left!!)
                }
                if (node.right != null) {
                    queue.add(node.right!!)
                }
            }
            result.add(level)
        }
        return result
    }

    fun levelOrder1(root: TreeNode?): List<List<Int>> {
        val result = ArrayList<ArrayList<Int>>()
        val queue: Queue<TreeNode?> = LinkedList<TreeNode?>()
        queue.add(root)
        while (!queue.isEmpty()) {
            val size = queue.size
            val sublist = ArrayList<Int>()
            for (i in 0 until size) {
                val node = queue.poll()
                if (node != null) {
                    sublist.add(node.`val`)
                    queue.offer(node.left)
                    queue.offer(node.right)
                }
            }
            if (sublist.isNotEmpty())
                result.add(sublist)
        }
        return result
    }

    //test
    fun main() {
        val root = TreeNode(3)
        root.left = TreeNode(9)
        root.right = TreeNode(20)
        root.right!!.left = TreeNode(15)
        root.right!!.right = TreeNode(7)
//        println(levelOrder(root))
        println(levelOrder1(root))
    }
}